engine/src/main/battlecode/world/GameWorld.java

```diff
@@ -264,49 +267,39 @@ public void removeSoup(MapLocation loc, int amount) {
     // ***********************************
 
     public int getPollution(MapLocation loc) {
-        if (pollutionNeedsUpdate)
-            calculatePollution();
-        return this.gameMap.onTheMap(loc) ? this.pollution[locationToIndex(loc)] : 0;
+        if (!this.gameMap.onTheMap(loc))
+            return 0;
+        int thisPollution = Math.round((globalPollution + this.localAdditivePollution[locationToIndex(loc)]) * this.localMultiplicativePollution[locationToIndex(loc)]);
+        return thisPollution;
     }
 
     public void addLocalPollution(int robotID, MapLocation loc, int radiusSquared, int additive, float multiplicative) {
         LocalPollutionEffect pE = new LocalPollutionEffect(loc, radiusSquared, additive, multiplicative);
         localPollutions.put(robotID, pE);
         getMatchMaker().addLocalPollution(loc, radiusSquared, additive, multiplicative);
-        pollutionNeedsUpdate = true;
+        // iterate over all locations affected by this robot
+        for (MapLocation affectedLoc : getAllLocationsWithinRadiusSquared(loc, radiusSquared)) {
+            this.localAdditivePollution[locationToIndex(affectedLoc)] += additive;
+            this.localMultiplicativePollution[locationToIndex(affectedLoc)] *= multiplicative;
+        }
     }
 
     public void resetPollutionForRobot(int robotID) {
+        if (!localPollutions.containsKey(robotID))
+            return;
+        // iterate over all locations affected by this robot
+        LocalPollutionEffect pE = localPollutions.get(robotID);
+        for (MapLocation affectedLoc : getAllLocationsWithinRadiusSquared(pE.loc, pE.radiusSquared)) {
+            this.localAdditivePollution[locationToIndex(affectedLoc)] -= pE.additiveEffect;
+            this.localMultiplicativePollution[locationToIndex(affectedLoc)] /= pE.multiplicativeEffect;
```

Comment:
anything preventing this from divmod 0?

Reply:
yeah, the multiplicative effect is always either 0.8 or 1.